### PR TITLE
Update to Trino

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

**Comment by findepi** [comment]
Would it be possible to split this PR into smaller ones?
Like 
- deps updates
- package renames & file moving aroun
- further renames, like server class, CLI, warnings, README (this can be subsplit, if grows too big)

**Comment by electrum** [comment]
@findepi This is still WIP and I've been rebasing it heavily, but I think it's mostly complete at this point and won't require changes to the first set of commits. I've tried to keep the commits small and logical, but given the scope of the change, it's not feasible to be "logically consistent" across the code base, nor have I tested individual commits after the code move portion.

I created #6486 with the set of pre commits. This should pass all the tests, and we can review/merge that now, while fixing the remaining tests in this PR.

**Review by dain — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1,92 +1,90 @@
...
-After building Presto for the first time, you can load the project into your IDE
-and run the server. We recommend using [IntelliJ
-IDEA](http://www.jetbrains.com/idea/). Because Presto is a standard Maven
-project, you can import it into your IDE using the root `pom.xml` file. In
-IntelliJ, choose Open Project from the Quick Start box or choose Open from the
-File menu and select the root `pom.xml` file.
+After building Trino for the first time, you can load the project into your IDE
+and run the server.  We recommend using
+[IntelliJ IDEA](http://www.jetbrains.com/idea/). Because Trino is a standard
+Maven project, you easily can import it into your IDE.  In IntelliJ, choose
```
Comment (findepi) [inline README.md:56]: is "easily" in the right place n this sentence?

File: README.md [hunk 2]
```diff
@@ -1,92 +1,90 @@
...
 
-Presto comes with sample configuration that should work out-of-the-box for
+### Running the full server
+
+Trino comes with sample configuration that should work out-of-the-box for
 development. Use the following options to create a run configuration:
 
 * Main Class: `io.trino.server.TrinoServer`
-* VM Options: `-ea -XX:+UseG1GC -XX:G1HeapRegionSize=32M -XX:+UseGCOverheadLimit -XX:+ExplicitGCInvokesConcurrent -Xmx2G -Dconfig=etc/config.properties -Dlog.levels-file=etc/log.properties -Djdk.attach.allowAttachSelf=true`
+* VM Options: `-ea -Dconfig=etc/config.properties -Dlog.levels-file=etc/log.properties -Djdk.attach.allowAttachSelf=true`
```
Comment (findepi) [inline README.md:81]: I see you removed redundant properties that are useful in general, like `-Xmx` (useful when you start a few JVMs on your laptop). In doing so, any reason to leave `-ea`, `-Djdk.attach.allowAttachSelf=true` or `-Dlog.levels-file` here? Are they not redundant in strict sense of the word?

File: README.md [hunk 3]
```diff
@@ -1,92 +1,90 @@
...
 IntelliJ, using `$MODULE_DIR$` accomplishes this automatically.
 
-Additionally, the Hive plugin must be configured with the location of your Hive
-metastore Thrift service. Add the following to the list of VM options, replacing
-`localhost:9083` with the correct host and port (or use the below value if you
-do not have a Hive metastore):
-
-    -Dhive.metastore.uri=thrift://localhost:9083
-
-### Using SOCKS for Hive or HDFS
```
Comment (findepi) [inline README.md:77]: Not that it was especially useful, but why remove?

File: client/trino-client/src/main/java/io/trino/client/PrestoHeaders.java [hunk 4]
```diff
@@ -24,24 +25,26 @@
...
+    public static final String PRESTO_ROLE = "X-Presto-Role";
+    public static final String PRESTO_PREPARED_STATEMENT = "X-Presto-Prepared-Statement";
+    public static final String PRESTO_TRANSACTION_ID = "X-Presto-Transaction-Id";
+    public static final String PRESTO_CLIENT_INFO = "X-Presto-Client-Info";
+    public static final String PRESTO_CLIENT_TAGS = "X-Presto-Client-Tags";
+    public static final String PRESTO_CLIENT_CAPABILITIES = "X-Presto-Client-Capabilities";
+    public static final String PRESTO_RESOURCE_ESTIMATE = "X-Presto-Resource-Estimate";
+    public static final String PRESTO_EXTRA_CREDENTIAL = "X-Presto-Extra-Credential";
+
+    // response headers
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/PrestoHeaders.java]: it looks like a line comment for next thing.

i find "section-like" comments generally not helpful (they tend to float away), but if you make one, please separate it from the next line or make it otherwise visually more catchy

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+        return prefix + "Clear-Transaction-Id";
+    }
+
+    public static ProtocolHeaders detectProtocol(Optional<String> alternateHeaderName, Set<String> headerNames)
+            throws ProtocolDetectionException
+    {
+        requireNonNull(alternateHeaderName, "alternateHeaderName is null");
+        requireNonNull(headerNames, "headerNames is null");
+
+        if (alternateHeaderName.isPresent() && !alternateHeaderName.get().equalsIgnoreCase("Trino")) {
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:192]: it should be illegal to set alternate header name to "Trino"

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            if (headerNames.contains("X-" + alternateHeaderName.get() + "-User")) {
+                if (headerNames.contains("X-Trino-User")) {
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:194]: This code assumes the user header (either `X-Trino-User` or `X-Alternate-User`) is always set, so it can use as the detection marker.
Thus, it should validate it is the case: if neither `X-Trino-User` or `X-Alternate-User` is set, it should throw `ProtocolDetectionException`

File: client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public final class ProtocolHeaders
+{
+    public static final ProtocolHeaders TRINO_HEADERS = new ProtocolHeaders("Trino");
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/ProtocolHeaders.java:24]: The value should be used in client code and tests and not used directly on the server side.
Would it be possible to move the value so that this enforcement is applied at compilation time?

File: core/trino-main/src/main/java/io/trino/server/security/InsecureAuthenticator.java [hunk 8]
```diff
@@ -53,11 +59,18 @@ public Identity authenticate(ContainerRequestContext request)
...
+            }
+            catch (ProtocolDetectionException e) {
+                // ignored
+                user = null;
+            }
         }
 
         if (user == null) {
-            throw new AuthenticationException("Basic authentication or " + PRESTO_USER + " must be sent", BasicAuthCredentials.AUTHENTICATE_HEADER);
+            throw new AuthenticationException("Basic authentication or " + TRINO_HEADERS.requestUser() + " must be sent", BasicAuthCredentials.AUTHENTICATE_HEADER);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/security/InsecureAuthenticator.java:73]: This should use detected protocol header, as the code above accesses `protocolHeaders.requestUser()` header.

(or, we change the message not to quote header name at all)

File: core/trino-main/src/main/java/io/trino/server/ProtocolConfig.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Optional;
+
+public class ProtocolConfig
+{
+    private String alternateHeaderName;
+
+    @Deprecated
+    public Optional<@Pattern(regexp = "[A-Za-z]+") String> getAlternateHeaderName()
+    {
+        return Optional.ofNullable(alternateHeaderName);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/ProtocolConfig.java:30]: You can declare the field as `Optional` directly, see `OAuth2Config` for example.

File: client/trino-client/src/test/java/io/trino/client/TestProtocolHeaders.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import com.google.common.collect.ImmutableSet;
+import org.testng.annotations.Test;
+
+import java.util.Optional;
+
+import static io.trino.client.ProtocolHeaders.detectProtocol;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.assertEquals;
+
+public class TestProtocolHeaders
```
Comment (findepi) [inline client/trino-client/src/test/java/io/trino/client/TestProtocolHeaders.java:25]: Ideally the feature should also have integration tests.
Since the goal is to make the old clients talk to current server, we could test just that.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java [hunk 11]
```diff
@@ -411,15 +407,12 @@ public void testTimeWithTimeZone()
                 assertEquals(rs.getObject(column), Time.valueOf(LocalTime.of(1, 39, 7))); // TODO this should represent TIME '09:39:07 +01:00'
                 assertThatThrownBy(() -> rs.getDate(column))
                         .isInstanceOf(SQLException.class)
-                        .hasMessage(serverSupportsVariablePrecisionTimeWithTimeZone()
-                                ? "Expected value to be a date but is: 09:39:07+01:00"
-                                : "Expected value to be a date but is: 09:39:07.000 +01:00");
+                        .hasMessage("Expected value to be a date but is: 09:39:07+01:00");
                 assertEquals(rs.getTime(column), Time.valueOf(LocalTime.of(1, 39, 7))); // TODO this should fail, or represent TIME '09:39:07'
+                // TODO (https://github.com/trinodb/trino/issues/5317) placement of precision parameter
```
Comment (findepi) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java:412]: Since the assertions are fluent and thus multi-step, this comment should be retained back on (or before) the line it pertains to.

File: client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java [hunk 12]
```diff
@@ -431,15 +424,12 @@ public void testTimeWithTimeZone()
                 assertEquals(rs.getObject(column), someBogusValue); // TODO this should represent TIME '01:39:07 +01:00'
                 assertThatThrownBy(() -> rs.getDate(column))
                         .isInstanceOf(SQLException.class)
-                        .hasMessage(serverSupportsVariablePrecisionTimeWithTimeZone()
-                                ? "Expected value to be a date but is: 01:39:07+01:00"
-                                : "Expected value to be a date but is: 01:39:07.000 +01:00");
+                        .hasMessage("Expected value to be a date but is: 01:39:07+01:00");
                 assertEquals(rs.getTime(column), someBogusValue); // TODO this should fail, or represent TIME '01:39:07'
+                // TODO (https://github.com/trinodb/trino/issues/5317) placement of precision parameter
```
Comment (findepi) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java:429]: same here & other places

File: testing/trino-test-jdbc-compatibility-old-driver/bin/run_tests.sh [hunk 13]
```diff
@@ -10,11 +10,14 @@ maven_run_tests="${maven} clean test -Dair.check.skip-all=true -Dmaven.javadoc.s
...
 current_version=$(${maven} help:evaluate -Dexpression=project.version -q -DforceStdout)
 previous_released_version=$((${current_version%-SNAPSHOT}-1))
-first_tested_version=312
+first_tested_version=351
 # test n-th version only
 version_step=3
 
 echo "Current version: ${current_version}"
+
+(( previous_released_version >= first_tested_version )) || exit 0
```
Comment (findepi) [inline testing/trino-test-jdbc-compatibility-old-driver/bin/run_tests.sh]: This should use explicit `if`. `||` for control flow does not add to readability

File: testing/trino-test-jdbc-compatibility-old-driver/src/test/java/io/trino/JdbcDriverCapabilities.java [hunk 14]
```diff
@@ -25,13 +26,28 @@ private JdbcDriverCapabilities() {}
...
     {
-        try (TrinoDriver driver = new TrinoDriver()) {
-            return driver.getMajorVersion();
+        return jdbcDriver().getMajorVersion();
+    }
+
+    public static Driver jdbcDriver()
+    {
+        return getDriver("jdbc:trino:")
+                .or(() -> getDriver("jdbc:presto:"))
```
Comment (findepi) [inline testing/trino-test-jdbc-compatibility-old-driver/src/test/java/io/trino/JdbcDriverCapabilities.java]: Why is that, if you removed testing for pore-351?

File: testing/trino-test-jdbc-compatibility-old-driver/src/test/java/io/trino/TestJdbcCompatibility.java [hunk 15]
```diff
@@ -104,18 +108,6 @@ public void tearDown()
         server.close();
     }
 
-    @Test
-    public void ensureUsingProperClientVersion()
```
Comment (findepi) [inline testing/trino-test-jdbc-compatibility-old-driver/src/test/java/io/trino/TestJdbcCompatibility.java]: Why was this removed?

File: testing/trino-test-jdbc-compatibility-old-server/pom.xml [hunk 16]
```diff
@@ -14,7 +14,6 @@
 
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
-        <dep.presto-jdbc-under-test>342-SNAPSHOT</dep.presto-jdbc-under-test>
```
Comment (findepi) [inline testing/trino-test-jdbc-compatibility-old-server/pom.xml]: Why was this removed and not renamed?

File: pom.xml [hunk 17]
```diff
@@ -1755,8 +1755,8 @@
         </profile>
 
         <profile>
-            <id>errorprone-compiler-presto</id><!-- airbase defines "errorprone-compiler" profile but set up this ourselves -->
-
+            <!-- airbase defines "errorprone-compiler" profile but set up this ourselves -->
+            <id>errorprone-compiler-trino</id>
```
Comment (findepi) [inline pom.xml]: Maybe we could remove the profile from airbase and have a simpler name here?

File: core/trino-main/src/main/java/io/trino/server/InternalAuthenticationManager.java [hunk 18]
```diff
@@ -43,7 +43,7 @@
 {
     private static final Logger log = Logger.get(InternalAuthenticationManager.class);
 
-    private static final String PRESTO_INTERNAL_BEARER = "X-Presto-Internal-Bearer";
+    private static final String TRINO_INTERNAL_BEARER = "X-Trino-Internal-Bearer";
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/InternalAuthenticationManager.java:46]: Should this go to `InternalHeaders` class?

File: core/trino-main/src/main/java/io/trino/server/Server.java [hunk 19]
```diff
@@ -68,12 +68,12 @@
 
 public class Server
 {
-    public final void start(String prestoVersion)
+    public final void start(String trinoVersion)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/Server.java:71]: `serverVersion` or just `version`.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java [hunk 20]
```diff
@@ -1048,7 +1048,7 @@ public synchronized void declareIntentionToWrite(ConnectorSession session, Write
         if (writeMode == WriteMode.DIRECT_TO_TARGET_EXISTING_DIRECTORY) {
             Map<List<String>, Action<PartitionAndMore>> partitionActionsOfTable = partitionActions.get(schemaTableName);
             if (partitionActionsOfTable != null && !partitionActionsOfTable.isEmpty()) {
-                throw new TrinoException(NOT_SUPPORTED, "Cannot insert into a table with a partition that has been modified in the same transaction when Presto is configured to skip temporary directories.");
+                throw new TrinoException(NOT_SUPPORTED, "Cannot insert into a table with a partition that has been modified in the same transaction when Trinoto is configured to skip temporary directories.");
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/SemiTransactionalHiveMetastore.java]: typo

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/util/TestHiveBucketing.java [hunk 21]
```diff
@@ -74,6 +74,10 @@
     @Test
     public void testHashingCompare()
     {
+        assertBucketEquals("string", "Trino rocks", 1132136730, -399107423);
+        assertEquals(HiveBucketing.getBucketNumber(1132136730, 4), 2);
+        assertEquals(HiveBucketing.getBucketNumber(-399107423, 4), 1);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/util/TestHiveBucketing.java]: These two are redundant as this is done by `assertBucketEquals` implicitly, isn't it?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/util/TestHiveBucketing.java [hunk 22]
```diff
@@ -74,6 +74,10 @@
     @Test
     public void testHashingCompare()
     {
+        assertBucketEquals("string", "Trino rocks", 1132136730, -399107423);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/util/TestHiveBucketing.java]: please move it before (or after) varchar case

File: plugin/trino-memsql/src/main/java/io/trino/plugin/memsql/MemSqlClient.java [hunk 23]
```diff
@@ -124,7 +124,7 @@ protected boolean filterSchema(String schemaName)
     }
 
     @Override
-    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    public Optional<ColumnMapping> toTrinoType(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
```
Comment (findepi) [inline plugin/trino-memsql/src/main/java/io/trino/plugin/memsql/MemSqlClient.java]: If we make a breaking change here, we should call the method `toColumnMapping`

File: plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java [hunk 24]
```diff
@@ -517,7 +517,7 @@ public void testTimestamp()
     public void testJson()
     {
         jsonTestCases(memSqlJsonDataType(value -> "JSON " + formatStringLiteral(value)))
-                .execute(getQueryRunner(), prestoCreateAsSelect("presto_test_json"));
+                .execute(getQueryRunner(), prestoCreateAsSelect("trino_test_json"));
```
Comment (findepi) [inline plugin/trino-memsql/src/test/java/io/trino/plugin/memsql/TestMemSqlTypeMapping.java]: we should use "test_json" in cases like that.

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlTypeMapping.java [hunk 25]
```diff
@@ -211,7 +211,7 @@ public void testReal()
                 .addRoundTrip("real", "nan()", REAL, "CAST(nan() AS real)")
                 .addRoundTrip("real", "-infinity()", REAL, "CAST(-infinity() AS real)")
                 .addRoundTrip("real", "+infinity()", REAL, "CAST(+infinity() AS real)")
-                .execute(getQueryRunner(), prestoCreateAsSelect("presto_test_real"));
+                .execute(getQueryRunner(), trinoCreateAsSelect("trino__test_real"));
```
Comment (findepi) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlTypeMapping.java]: doubled underscore, probably not intentional

**Review by findepi — COMMENTED** [review]
(skimming)

File: plugin/trino-memsql/src/main/java/io/trino/plugin/memsql/MemSqlClient.java (hunk 23 above)
Comment (findepi) [reply plugin/trino-memsql/src/main/java/io/trino/plugin/memsql/MemSqlClient.java]: https://github.com/trinodb/trino/pull/6500
